The instance passed to the [`lock`
statement](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock) should be a dedicated private field.

## Why is this an issue?

If the instance representing an exclusively acquired lock is publicly accessible, another thread in another part of the program could accidentally
attempt to acquire the same lock. This increases the likelihood of [deadlocks](https://en.wikipedia.org/wiki/Deadlock).

For example, a `string` should never be used for locking. When a `string` is [interned](https://en.wikipedia.org/wiki/Interning_%28computer_science%29) by the runtime, it can be shared by multiple threads, breaking the
locking mechanism.

Instead, a dedicated private [`Lock`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.lock?view=net-9.0) object
instance (or `object` instance, for frameworks before .Net 9) should be used for locking. This minimizes access to the lock instance and
therefore prevents accidential lock sharing.

The following objects are considered potentially prone to accidental lock sharing:

- a reference to [this](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/this): if the instance is publicly
  accessibly, the lock might be shared
- a [Type](https://learn.microsoft.com/en-us/dotnet/api/system.type) object: if the type class is publicly accessibly, the lock might
  be shared
- a [string](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/) literal or instance: if any other part of the
  program uses the same string, the lock is shared because of interning

## How to fix it

### Code examples

#### Noncompliant code example

    void MyLockingMethod()
    {
        lock (this) // Noncompliant
        {
            // ...
        }
    }

#### Compliant solution

    #if NET9_0_OR_GREATER
    private readonly Lock lockObj = new();
    #else
    private readonly object lockObj = new();
    #endif
    
    void MyLockingMethod()
    {
        lock (lockObj)
        {
            // ...
        }
    }

## Resources

### Documentation

- Wikipedia - [Thread](https://en.wikipedia.org/wiki/Thread_%28computing%29)
- Wikipedia - [Locking](https://en.wikipedia.org/wiki/Lock_%28computer_science%29)
- Wikipedia - [Deadlock](https://en.wikipedia.org/wiki/Deadlock)
- Wikipedia - [Interning](https://en.wikipedia.org/wiki/Interning_%28computer_science%29)
- Microsoft Learn - [String interning by the runtime](https://learn.microsoft.com/en-us/dotnet/api/system.string.intern#remarks)
- Microsoft Learn - [Managed Threading Best
  Practices](https://docs.microsoft.com/en-us/dotnet/standard/threading/managed-threading-best-practices)
- Microsoft Learn - [The lock statement - ensure
  exclusive access to a shared resource](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock)